runtime/include/cloe/data_broker.hpp

```diff
@@ -0,0 +1,1099 @@
+ *     An optional container storing the value of a signal and abstracting the interaction with Signal.
+ *
+ * Background:
+ *   Real world simulations utilize manifolds of signals (variables) and coresponding types.
+ *   Still the very basic problem around a single signal (variable) boils down to CRUD
+ *   - declaring a signal (Create)
+ *   - reading its value / receiving value-changed events (Read)
+ *   - writing its value / triggering value-changed events (Update)
+ *   - NOT NEEDED: undeclaring a signal (Delete)
+ *
+ * Concept:
+ *   Abstract CRU(D) operations by registering signals in a uniform way by their name.
+ */
+
+#pragma once
+#ifndef CLOE_DATA_BROKER_HPP_
+#define CLOE_DATA_BROKER_HPP_
+
+#include <any>
+#include <cassert>
+#include <functional>
+#include <map>
+#include <memory>
+#include <regex>
+#include <stdexcept>
+#include <string_view>
+#include <vector>
+
+#include <fmt/format.h>
+
+namespace cloe {
+
+namespace databroker {
+
+/**
+ * Maps a predicate to type T or type int.
+ * \returns T, if condition is true
+ *          int, otherwise
+ */
+template <typename T, bool condition>
+struct type_t_or_int_if {
+  using type = T;
+};
+
+template <typename T>
+struct type_t_or_int_if<T, false> {
+  using type = int;
+};
+
+/**
+ * Predicate which determines whether a type is incompatible with the data
+ * broker.
+ */
+template <typename T>
+constexpr bool is_incompatible_type_v = std::is_void_v<T> || std::is_reference_v<T>;
+
+/**
+ * Determines a datatype which is compatible with the data broker, derived from
+ * the template type.
+ *
+ * \returns T, if T is compatible with the data broker;
+ *          int, otherwise
+ * \note Purpose is to suppress irrelevant compiler errors, in case of
+ *       incompatible data types
+ */
+template <typename T>
+using compatible_base_t = typename type_t_or_int_if<T, !is_incompatible_type_v<T>>::type;
+
+/**
+ * Argument-/Return-Type of signal related functions
+ */
+template <typename T>
+using signal_type_cref_t = databroker::compatible_base_t<T> const&;
+
+/**
+ * Type of event function, which is called when the value of a signal changed
+ */
+template <typename T>
+using on_value_changed_callback_t = std::function<void(signal_type_cref_t<T>)>;
+
+/**
+ * Abstract event implementation.
+ *
+ * \tparam TArgs Arguments of the event handler function
+ * \note: Design-Goals:
+ * - Design-#1: Unsubscribing from an event is not intended
+ */
+template <typename... TArgs>
+class Event {
+ public:
+  using eventhandler = std::function<void(TArgs...)>;
+  using eventhandler_collection = std::vector<eventhandler>;
+
+ private:
+  eventhandler_collection eventhandlers_{};
+
+ public:
+  /**
+   * Add an event handler to this event.
+   */
+  void add(eventhandler eventhandler) { eventhandlers_.emplace_back(std::move(eventhandler)); }
+
+  /**
+   * Return the number of event handlers subscribed to this event.
+   */
+  auto count() const { return eventhandlers_.size(); }
```

Comment:
Use `std::size_t` since this:
- is consistent with the rest
- will not change
- makes the interface explicit
- does not depend on implementation detail
- improves readability

Reply:
If we want to make more use of auto, there's an argument to be made for that as well. I'm a little on the fence.